File: src/main/java/dev/liquid/osamaclient/feature/FeatureManager.java
package dev.liquid.osamaclient.feature;

import dev.liquid.osamaclient.feature.impl.esp.ParticleESP;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class FeatureManager {
  public static FeatureManager instance;
  private final ArrayList<IFeature> features = new ArrayList<>();

  public static FeatureManager getInstance() {
    if (instance == null) {
      instance = new FeatureManager();
    }
    return instance;
  }

  public ArrayList<IFeature> loadFeatures() {
    List<IFeature> features = Arrays.asList(
        ParticleESP.getInstance()
    );
    this.features.addAll(features);
    return this.features;
  }

  public void disableFeatures(boolean disablePassiveFeatures){
    this.features.forEach(feature -> {
      if (!feature.isPassiveFeature() || (feature.isPassiveFeature() && disablePassiveFeatures)) {
        feature.disable();
      }
    });
  }
}

File: src/main/kotlin/dev/liquid/osamaclient/feature/FeatureManager.kt
package dev.liquid.osamaclient.feature

import dev.liquid.osamaclient.feature.implementation.esp.ParticleESP
import java.util.*
import java.util.function.Consumer

class FeatureManager {
  // Singleton Getter
  companion object {
    private var instance: FeatureManager? = null
    fun getInstance(): FeatureManager {
      if (instance == null) {
        instance = FeatureManager()
      }
      return instance!!
    }
  }

  private val features = mutableListOf<IFeature>()
  fun loadFeatures(): List<IFeature> {
    val features = listOf<IFeature>(
      ParticleESP.getInstance(),
    )
    this.features.addAll(features)
    return this.features
  }
  fun disableFeatures(disablePassiveFeatures: Boolean) {
    features.forEach(Consumer { feature: IFeature ->
      if (!feature.isPassiveFeature || feature.isPassiveFeature && disablePassiveFeatures) {
        feature.disable()
      }
    })
  }
}

